src/confchange/changer.rs

```diff
@@ -0,0 +1,361 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crate::eraftpb::{ConfChangeSingle, ConfChangeType};
+use crate::tracker::{Configuration, ProgressMap, ProgressTracker};
+use crate::{Error, Result};
+
+/// Change log for progress map.
+pub enum MapChangeType {
+    Add,
+    Remove,
+}
+
+/// Changes made by `Changer`.
+pub type MapChange = Vec<(u64, MapChangeType)>;
+
+/// A map that stores updates instead of apply them directly.
+pub struct IncrChangeMap<'a> {
+    changes: MapChange,
+    base: &'a ProgressMap,
+}
+
+impl IncrChangeMap<'_> {
+    pub fn into_changes(self) -> MapChange {
+        self.changes
+    }
+
+    fn contains(&self, id: u64) -> bool {
+        match self.changes.iter().rfind(|(i, _)| *i == id) {
+            Some((_, MapChangeType::Remove)) => false,
+            Some((_, MapChangeType::Add)) => true,
+            None => self.base.contains_key(&id),
+        }
+    }
+}
+
+/// Changer facilitates configuration changes. It exposes methods to handle
+/// simple and joint consensus while performing the proper validation that allows
+/// refusing invalid configuration changes before they affect the active
+/// configuration.
+pub struct Changer<'a> {
+    tracker: &'a ProgressTracker,
+}
+
+impl Changer<'_> {
+    /// Creates a changer.
+    pub fn new(tracker: &ProgressTracker) -> Changer {
+        Changer { tracker }
+    }
+
+    /// Verifies that the outgoing (=right) majority config of the joint
+    /// config is empty and initializes it with a copy of the incoming (=left)
+    /// majority config. That is, it transitions from
+    /// ```text
+    ///     (1 2 3)&&()
+    /// ```
+    /// to
+    /// ```text
+    ///     (1 2 3)&&(1 2 3)
+    /// ```.
+    ///
+    /// The supplied changes are then applied to the incoming majority config,
+    /// resulting in a joint configuration that in terms of the Raft thesis[1]
+    /// (Section 4.3) corresponds to `C_{new,old}`.
+    ///
+    /// [1]: https://github.com/ongardie/dissertation/blob/master/online-trim.pdf
+    pub fn enter_joint(
+        &self,
+        auto_leave: bool,
+        ccs: &[ConfChangeSingle],
+    ) -> Result<(Configuration, MapChange)> {
+        if joint(self.tracker.conf()) {
+            return Err(Error::ConfChangeError(
+                "configuration is already joint".to_owned(),
+            ));
+        }
+        let (mut cfg, mut prs) = self.check_and_copy()?;
+        if cfg.voters().incoming.is_empty() {
+            // We allow adding nodes to an empty config for convenience (testing and
+            // bootstrap), but you can't enter a joint state.
+            return Err(Error::ConfChangeError(
+                "can't make a zero-voter config joint".to_owned(),
+            ));
+        }
+        cfg.voters
+            .outgoing
+            .extend(cfg.voters.incoming.iter().cloned());
```

Comment:
The check at L71 ensures that outgoing is empty. Assigning new empty map is necessary in etcd, because it uses nil map which can't be modified. But we use empty map that doesn't have the issue.